tests/common/src/lib.rs

```diff
@@ -16,169 +16,395 @@
-    mut runner: MemoryRunner,
-    code: Vec<u8>,
-    requests: Vec<Req>,
-) -> Vec<Rep> {
-    runner
-        .init_program(InitializeProgramInfo {
-            new_program_id: 1.into(),
-            source_id: 0.into(),
-            code,
-            message: ExtMessage {
-                id: 1000001.into(),
-                payload: "init".as_bytes().to_vec(),
-                gas_limit: u64::MAX,
-                value: 0,
-            },
-        })
-        .expect("failed to init program");
-
-    let mut nonce = 0;
-
-    let mut data: Vec<(u64, MessageId, Option<Rep>)> = Vec::new();
-
-    for request in requests {
-        let message_id: MessageId = (nonce + 1000002).into();
-        data.push((nonce, message_id, None));
-        runner.queue_message(MessageDispatch {
-            source_id: 0.into(),
-            destination_id: 1.into(),
-            data: ExtMessage {
-                id: message_id,
-                gas_limit: u64::MAX,
-                value: 0,
-                payload: request.encode(),
-            },
-        });
-        nonce += 1;
-    }
-
-    while runner.run_next(u64::MAX).handled != 0 {}
-
-    let Storage { log, .. } = runner.complete();
-
-    assert_eq!(
-        log.get().first().map(|m| m.payload().to_vec()),
-        Some(b"CREATED".to_vec())
-    );
-
-    for message in log.get().iter() {
-        for (_, search_message_id, ref mut reply) in data.iter_mut() {
-            if message
-                .reply
-                .map(|(msg_id, _)| msg_id == *search_message_id)
-                .unwrap_or(false)
-            {
-                *reply = Some(
-                    Rep::decode(&mut message.payload.as_ref()).expect("Failed to decode reply"),
-                );
-            }
+pub struct InitProgram<E: Encode> {
+    pub program_id: Option<ProgramId>,
+    pub source_id: Option<ProgramId>,
+    pub code: Vec<u8>,
+    pub message: Option<Message<E>>,
+}
+
+impl<E: Encode> InitProgram<E> {
+    pub fn id<P: Into<ProgramId>>(self, id: P) -> Self {
+        Self {
+            program_id: Some(id.into()),
+            ..self
         }
     }
 
-    data.into_iter()
-        .map(|(_, _, reply)| reply.expect("No reply for message"))
-        .collect()
+    pub fn source_id<P: Into<ProgramId>>(self, id: P) -> Self {
+        Self {
+            source_id: Some(id.into()),
+            ..self
+        }
+    }
+
+    pub fn message<M: Into<Message<E>>>(self, message: M) -> Self {
+        Self {
+            message: Some(message.into()),
+            ..self
+        }
+    }
 }
 
-pub struct MessageData<P: Encode> {
-    pub id: MessageId,
-    pub payload: P,
-    pub gas_limit: u64,
-    pub value: u128,
+impl<C, E> From<C> for InitProgram<E>
+where
+    E: Encode,
+    C: Into<Vec<u8>>,
+{
+    fn from(code: C) -> Self {
+        Self {
+            program_id: None,
+            source_id: None,
+            code: code.into(),
+            message: None,
+        }
```

Comment:
Although this syntax is nice, I would prefer not to use this pattern. Unless Default implementation is const, this pattern forces to build entire default structure before initializing the fields. This could be easily misused in case of costly initialization of already provided fields. I do understand that this is not the case here, I would only use this syntax with already initialized structures anyway.

Reply:
Ok, agree.